### DISCUSSION TITLE
Can I use "torch.autograd.grad" in "training_step" funtion?

### DISCUSSION THREAD

**Discussion by chenshuang-zhang** [Lightning-AI/pytorch-lightning#10731] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
Hello! 

I hope to **calculate the  gradient of the loss on the model input** every batch when training. The calculated gradients are then processed by some other functions and added into the loss function. The way I do this  in pytorch is like this, with the `torch.autograd.grad` function:

```python
model_input.requires_grad = True
model.zero_grad()
model.eval()
grads = torch.autograd.grad(loss, model_input, grad_outputs=None, only_inputs=True, retain_graph=False)[0]
model.train()
```

Can I add these codes directly into the `training_step` function in pytorch-lightning? 

My concern is: 
1. Will this gradient calculation (`torch.autograd.grad` function)  influence the accuracy of model training since I add it in the `training_step` function? 
2. Do I need to set `model.zero_grad()`,   `model.eval()`  and `model.train()`  when calculating the gradients on the input?

Thank you very much!

**Comment by akihironitta** [comment, marked as answer]
> 1. Will this gradient calculation (`torch.autograd.grad` function)  influence the accuracy of model training since I add it in the `training_step` function?

Yes, I believe. I don't think it'll work with `retain_graph=False` there because backward pass uses the graph to compute gradients wrt weights using the `loss` returned from `training_step`.

> 2. Do I need to set `model.zero_grad()`,   `model.eval()`  and `model.train()`  when calculating the gradients on the input?

Yes, partially. I think you need `zero_grad()` after `grads = torch.autograd.grad(...)` to avoid accumulating gradients wrt weights, but I'm not sure why you need `eval()` and `train()`.

**Reply by tchaton** [reply]
Hey @YingYingFight,

1. You can call autograd yourself, but you need to be aware this doesn't work with DDP nor precision 16. As an alternative, you can use `automatic_optimization=False`, use self.manual_backward(loss, ...) to compute the gradients and access them by doing grads = [p.grad for p in self.paramters()]

2. If you use manual optimization, you would have to call zero_grad yourself.

**Reply by chenshuang-zhang** [reply]
> 

Hello，@akihironitta. Thank you very much for your help! 

For the first question, I still feel a little confused.  I calculate the gradient of the loss on the model **_input_**.  Do you mean it still influence the **_weight update_**, and thus influence the model accuracy?

Thank you!

**Reply by chenshuang-zhang** [reply]
> 
Hello, @tchaton. Thank you very much for your help! I will try using the `self.manual_backward` in the code! However, I'm still confused about your first comment. 


`You can call autograd yourself, but you need to be aware this works as with DDP nor precision 16.`
**Do you mean if I use autograd, the DDP and precision 16 will be invalid?**

Thank you!

**Reply by akihironitta** [reply]
Hi @YingYingFight

> Do you mean it still influence the weight update, and thus influence the model accuracy?

Correction! I had a look at your code again and realised that you set `only_inputs=True`, so it shouldn't impact on anything.

FYI, `only_inputs` is deprecated since `torch>=1.10` so if you are using 1.10 or later, you can just get rid of it :)
> only_inputs argument is deprecated and is ignored now (defaults to True). To accumulate gradient for other parts of the graph, please use torch.autograd.backward.
> https://pytorch.org/docs/stable/generated/torch.autograd.grad.html#torch.autograd.grad

I'm sorry for any inconvenience, and thank you for asking back!

---

For manual optimization, see https://pytorch-lightning.readthedocs.io/en/latest/common/optimizers.html#manual-optimization :)

**Reply by chenshuang-zhang** [reply]
> Hi @YingYingFight
> 
> > Do you mean it still influence the weight update, and thus influence the model accuracy?
> 
> Correction! I had a look at your code again and realised that you set `only_inputs=True`, so it shouldn't impact on anything.
> 
> FYI, `only_inputs` is deprecated since `torch>=1.10` so if you are using 1.10 or later, you can just get rid of it :)
> 
> > only_inputs argument is deprecated and is ignored now (defaults to True). To accumulate gradient for other parts of the graph, please use torch.autograd.backward.
> > https://pytorch.org/docs/stable/generated/torch.autograd.grad.html#torch.autograd.grad
> 
> I'm sorry for any inconvenience, and thank you for asking back!
> 
> For manual optimization, see https://pytorch-lightning.readthedocs.io/en/latest/common/optimizers.html#manual-optimization :)

@akihironitta,   Thank you very much for your help! Could you please help me check if I understand this problem right?   My understandings  are as follows:

Aim:     (1) Calculate the gradient of the loss on the **_model input_** every batch when **_training_**.
             (2) The calculated gradients are then processed by some other functions and added into the loss function. 

Method:  (1) In `training_step` function,   use `torch.autograd.grad`  to calculate the gradient of loss on _**model input.**_
                (2) In  `training_step` function,  use manual optimization   (like `self.manual_backward`)  to calculate the loss.  [https://pytorch-lightning.readthedocs.io/en/latest/common/optimizers.html#manual-optimization](url)

Influence: (1) If I use the original loss,   the `torch.autograd.grad`  with `torch.autograd.grad`  will not impact anything.
                 (2) If I hope to obtain the second aim, by manual optimization,  the model will be optimized with my new loss.

**Do I understand it right? Thank you!**

**Reply by akihironitta** [reply]
@YingYingFight I guess you're right. As Thomas mentioned, simply implementing it with automatic optimization, it won't work with DDP nor mixed precision training, so I'd suggest you use manual optimization.

**Reply by chenshuang-zhang** [reply]
@akihironitta Thank you for your help! I'm new to PyTorchLightning, so I'm not sure if I understand your comment right. 
So the conclusion is:

(1) If I use `torch.autograd.grad` in the `training_step` function, and the _**automatic optimization**_, the `torch.autograd.grad` _**won't**_ work with DDP nor mixed precision training. 

(2) With _**automatic optimization**_,  the `torch.autograd.grad` **_will_** work with DDP and  mixed precision training. 

Do I understand it right? Thank you!

**Reply by tchaton** [reply]
Hey @YingYingFight,

When using DDP, each process will compute its own gradients based on the local data. Therefore, you can compute the local gradients but you need to be careful when updating the model weights, as this needs to be done using the global gradients = MEAN(all local gradients) through a all_reduce.

**Reply by chenshuang-zhang** [reply]
@tchaton Thank you very much for your detailed explanations! I will try as you said!

**Reply by chenshuang-zhang** [reply]
> Hey @YingYingFight,
> 
> When using DDP, each process will compute its own gradients based on the local data. Therefore, you can compute the local gradients but you need to be careful when updating the model weights, as this needs to be done using the global gradients = MEAN(all local gradients) through a all_reduce.

Hello @tchaton.  Sorry I have another two questions about your comment. 

(1) If I use only one GPU when training and no mixed precision training,  can I use `torch.autograd.grad` in  `automatic optimization`? 

(2) Can I use `model.zero_grad()` in the `training_step` function  in  `automatic optimization`? 

Thank you very much!

**Reply by tchaton** [reply]
(1) Yes.

(2) With Automatic Optimization, Lightning takes care of zero_grad. However, with manual, you can use model.zero_grad() too ;)

**Reply by chenshuang-zhang** [reply]
> (1) Yes.
> 
> (2) With Automatic Optimization, Lightning takes care of zero_grad. However, with manual, you can use model.zero_grad() too ;)

Hi, @tchaton. It's so kind of you!  I tried the manual optimization these days, and got what I want already!  

To make sure I use correctly,  can you help me with the following question?  

 **_With manual optimization_**,  do I need to add some codes about **_mixed precision training_**？ 

Thank you very much!

**Reply by tchaton** [reply]
For, manual optimization, self.manual_backward takes care of precision :)

**Reply by chenshuang-zhang** [reply]
> For, manual optimization, self.manual_backward takes care of precision :)

Cool! Thank you!
